packages/teko/src/Tpetra/Teko_BlockingTpetra.cpp

```diff
@@ -164,148 +165,193 @@ void many2one(Tpetra::MultiVector<ST, LO, GO, NT>& one,
-
-  const size_t invalid = Teuchos::OrdinalTraits<size_t>::invalid();
-
-  // insert each row into subblock
-  // Count the number of entries per row in the new matrix
-  for (LO localRow = 0; localRow < numMyRows; localRow++) {
-    size_t numEntries = invalid;
-    GO globalRow      = gRowMap->getGlobalElement(localRow);
-    LO lid            = A->getRowMap()->getLocalElement(globalRow);
-    TEUCHOS_ASSERT(lid > -1);
-
-    A->getLocalRowCopy(lid, indices, values, numEntries);
-
-    LO numOwnedCols = 0;
-    auto data       = local2ContigGIDs.getLocalViewHost(Tpetra::Access::ReadOnly);
-    for (size_t localCol = 0; localCol < numEntries; localCol++) {
-      TEUCHOS_ASSERT(indices(localCol) > -1);
-
-      // if global id is not owned by this column
-
-      GO gid = data(indices[localCol], 0);
-      if (gid == -1) continue;  // in contiguous row
-      numOwnedCols++;
-    }
-    nEntriesPerRow[localRow] = numOwnedCols;
+  LO numMyRows = rangeMap->getLocalNumElements();
+
+  auto nEntriesPerRow =
+      Kokkos::View<LO*>(Kokkos::ViewAllocateWithoutInitializing("nEntriesPerRow"), numMyRows);
+  auto prefixSumEntriesPerRow = Kokkos::View<LO*>(
+      Kokkos::ViewAllocateWithoutInitializing("prefixSumEntriesPerRow"), numMyRows + 1);
+
+  const auto invalid = Teuchos::OrdinalTraits<GO>::invalid();
+
+  LO totalNumOwnedCols = 0;
+  auto A_dev           = A->getLocalMatrixDevice();
+  auto gRowMap_dev     = gRowMap->getLocalMap();
+  auto A_rowmap_dev    = A->getRowMap()->getLocalMap();
+  auto data            = local2ContigGIDs.getLocalViewDevice(Tpetra::Access::ReadOnly);
+
+  using matrix_execution_space =
+      typename Tpetra::CrsMatrix<ST, LO, GO, NT>::local_matrix_device_type::execution_space;
+  using team_policy = Kokkos::TeamPolicy<matrix_execution_space>;
+  using team_member = typename team_policy::member_type;
+
+  LO maxNumEntriesSubblock = 0;
+  Kokkos::parallel_reduce(
+      Kokkos::RangePolicy<Kokkos::Schedule<Kokkos::Dynamic>, matrix_execution_space>(0, numMyRows),
+      KOKKOS_LAMBDA(const LO localRow, LO& sumNumEntries, LO& maxNumEntries) {
+        GO globalRow             = gRowMap_dev.getGlobalElement(localRow);
+        LO lid                   = A_rowmap_dev.getLocalElement(globalRow);
+        const auto sparseRowView = A_dev.row(lid);
+
+        LO numOwnedCols = 0;
+        for (auto localCol = 0; localCol < sparseRowView.length; localCol++) {
+          GO gid = data(sparseRowView.colidx(localCol), 0);
+          if (gid == invalid) continue;
+          numOwnedCols++;
+        }
+        nEntriesPerRow(localRow) = numOwnedCols;
+        sumNumEntries += numOwnedCols;
+        maxNumEntries = Kokkos::max(maxNumEntries, numOwnedCols);
+      },
+      totalNumOwnedCols, Kokkos::Max<LO>(maxNumEntriesSubblock));
+
+  auto numPerRowHost = Kokkos::create_mirror_view(nEntriesPerRow);
+  Kokkos::deep_copy(numPerRowHost, nEntriesPerRow);
+
+  auto prefixSumHost = Kokkos::create_mirror_view(prefixSumEntriesPerRow);
```

Comment:
It might be possible to merge the summation into the parallel_reduce above.

Reply:
I think you're right if I switch the `parallel_reduce` above to a `parallel_scan`.

Reply:
Done.